vllm/distributed/kv_transfer/kv_connector/v1/shared_storage_connector.py

```diff
@@ -0,0 +1,364 @@
+                    filename)["kv_cache"].cuda()
+                inject_kv_into_layer(kv_cache_layer, kv_cache,
+                                     request.slot_mapping)
+
+    def wait_for_layer_load(self, layer_name: str) -> None:
+        """Blocking until the KV for a specific layer is loaded into vLLM's
+        paged buffer. 
+        
+        This interface will be useful for layer-by-layer pipelining.
+
+        Args:
+            layer_name: the name of that layer
+        """
+        return
+
+    def save_kv_layer(self, layer_name: str, kv_layer: torch.Tensor,
+                      attn_metadata: "AttentionMetadata", **kwargs) -> None:
+        """Start saving the a layer of KV cache from vLLM's paged buffer 
+        to the connector.
+
+        Args:
+            layer_name (str): the name of the layer.
+            kv_layer (torch.Tensor): the paged KV buffer of the current 
+                layer in vLLM.
+            attn_metadata (AttentionMetadata): the attention metadata.
+            **kwargs: additional arguments for the save operation.
+        """
+
+        def extract_kv_from_layer(
+            layer: torch.Tensor,
+            slot_mapping: torch.Tensor,
+        ) -> torch.Tensor:
+            """Extract the KV cache from the layer.
+
+            Assume the shape of the layer is (2, num_pages, page_size, xxx).
+            """
+            num_pages, page_size = layer.shape[1], layer.shape[2]
+            return layer.reshape(2, num_pages * page_size, -1)[:, slot_mapping,
+                                                               ...]
+
+        connector_metadata = self._get_connector_metadata()
+        assert isinstance(connector_metadata, SharedStorageConnectorMetadata)
+        for request in connector_metadata.requests:
+            if request.is_store:
+                filename = self.generate_filename_debug(
+                    layer_name, request.token_ids)
+                kv_cache = extract_kv_from_layer(kv_layer,
+                                                 request.slot_mapping)
+                tensors = {"kv_cache": kv_cache.cpu().detach()}
+                safetensors.torch.save_file(tensors, filename)
+
+    def wait_for_save(self):
+        return
+
+    def get_external_prefix_cache_blocks(
+        self,
+        request: "Request",
+        computed_blocks: list["KVCacheBlock"],
+        num_computed_tokens: int,
+        kv_cache_manager: "KVCacheManager",
+    ) -> list["KVCacheBlock"]:
+        """Get the external prefix cache blocks from the connector.
+
+        This function may change the state of the connector, which will be 
+        used by `attach_connector_meta` later.
+
+        Args:
+            request (Request): the request object.
+            computed_blocks (list[KVCacheBlock]): the 'local' computed blocks.
+            num_computed_tokens (int): the number of 'local' computed tokens.
+            kv_cache_manager (KVCacheManager): the KV cache manager to 
```

Comment:
It is a bit confusing to include blocks allocation/free under the semantic of "get_". Are we trying to allocate vLLM's paged KV cache here for external prefix cache? Maybe we should update the function into "get_and_allocate". 

Also is this the right place to add heuristic like whether we want to recompute instead of fetching remote cache?